### ISSUE TITLE
(iOS 8 GM iPhone5C) TypeError: Attempted to assign to readonly property

### ISSUE THREAD

**Issue by HAKASHUN** [angular/angular.js#9128]
I am using on iOS 8 GM(iPhone5C).
With this device, I am getting this error in this line(https://github.com/angular/angular.js/blob/master/src/ng/rootScope.js#L216)
But on iOS 8 GM(iPhone5S) and iOS 7.x (iPhone5C, iPhone5S), this error does not occur.

To resolve this error, I rewrote the code as follows:
``` javascript
$new: function(isolate) {
        var child;

        if (isolate) {
          child = new Scope();
          child.$root = this.$root;
          // ensure that there is just one async queue per $rootScope and its children
          child.$$asyncQueue = this.$$asyncQueue;
          child.$$postDigestQueue = this.$$postDigestQueue;
        } else {
          // Only create a child scope class if somebody asks for one,
          // but cache it to allow the VM to optimize lookups.
          if (!this.$$ChildScope) {
            this.$$ChildScope = function ChildScope() {
              this['$$watchers'] = this['$$nextSibling'] =
                  this['$$childHead'] = this['$$childTail'] = null;
              this['$$listeners'] = {};
              this['$$listenerCount'] = {};
              this['$id'] = nextUid();
              this['$$ChildScope'] = null;
            };
            this.$$ChildScope.prototype = this;
          }
          child = new this.$$ChildScope();
        }
        child['this'] = child;
        child['$parent'] = this;
        child['$$prevSibling'] = this.$$childTail;
        if (this.$$childHead) {
          this.$$childTail.$$nextSibling = child;
          this.$$childTail = child;
        } else {
          this.$$childHead = this.$$childTail = child;
        }
        return child;
      },
```

The situation has improved, but I do not understand the reason...

**Comment by edzis** [comment]
This seems to be an iOS 8 bug - http://stackoverflow.com/a/27401601

On bugsnag error logs for my application I also see that only on Mobile safari with iOS 8.0 and 8.1

Below is a stack trace for angular 1.2.28. The same issue is there also for 1.2.24.

```
TypeError Attempted to assign to readonly property.
    angular.js:12188:37 $$childScopeClass
    angular.js:12198:8 prototype
    angular.js:6161:29 
    angular.js:6773:35 controllersBoundTransclude
    angular.js:21409:16 link
    angular.js:6752:19 nodeLinkFn
    angular.js:6146:23 compositeLinkFn
    angular.js:6746:35 nodeLinkFn
    angular.js:6954:36 
    angular.js:8171:13 
    angular.js:11682:81 wrappedCallback
    angular.js:11768:34 
    angular.js:12811:28 prototype
    angular.js:12623:36 prototype
    angular.js:12915:12 prototype
    angular.js:19264:29 
    angular.js:2853:14 
    angular.js:325:22 forEach
    angular.js:2852:12 eventHandler
```

**Comment by elmalto** [comment]
Is there any other fix for this besides removing 'use strict'? I would like to CDN angular and this would disable it...

**Comment by DropsOfSerenity** [comment]
Monkey patching worked for me too, Obviously not a good solution, and this is a bug that is breaking angular out in production. Looks like ios8 webkit really dislikes assignment of .$$variables

**Comment by Nivani** [comment]
Making some changes from dot notation to brackets notation by monkey patching like @jhunken did worked for me too. Very strange problem.

**Comment by m3kka** [comment]
Still no word about an official solution?

**Comment by rjbernaldo** [comment]
Removing use strict did it for me. Still waiting on the official solution though.

**Comment by hamstu** [comment]
@rjbernaldo To clarify, are you removing 'use strict' from angular.js, your own source files, or both?

**Comment by rjbernaldo** [comment]
@hamstu I removed it from angular.js only.

**Comment by booleanbetrayal** [comment]
Removing `use strict;` fixed the issue on my end as well. I'm fairly convinced Apple is trying to kill off web-apps with breakage like this.  =]

**Comment by pascalc** [comment]
Removing `'use strict';` did stop the crashes but seems to have caused another problem: now my app seems to hang instead of crashing, and the tab in Safari becomes totally unresponsive. This even happens in the iOS simulator, where the simulator process starts using 200% CPU on my mac.

**Comment by caitp** [comment]
could you post a stack trace of the crash, along with versions of al the libraries you're using? that would be helpfulp.

**Comment by caitp** [comment]
oh hang on, I see what it is... wow I don't think that's the right behaviour in JSC, but just to check...

**Comment by caitp** [comment]
I haven't been able to reproduce this on my iphone, ipad, or any of the ios simulators I've tried. It looks to me like the issue is the `prototype` property being incorrectly defined as non-writable in JSC, which would be a bug (and, would explain why it's no longer reproducible if they've fixed it). If that's not the bug being hit, then please supply a reproduction for us.

**Comment by m3kka** [comment]
To me it seems more like that the bug creating this problem is https://bugs.webkit.org/show_bug.cgi?id=138038 (posted in a previous answer by @edzis)

**Comment by caitp** [comment]
@m3kka we aren't using O.create anywhere in 1.2.x

**Comment by m3kka** [comment]
In our situation this problem started when we moved  to 1.3.x branch, in which source code I can find some instances of Object.create() 

Am I completely off track?

**Comment by caitp** [comment]
The original bug report (and most of the comments) indicate that this affects version 1.2.x

**Comment by caitp** [comment]
at any rate, anyone experiencing this issue, provide a reproduction :<

**Comment by pascalc** [comment]
Here's what I get out of the Safari console from the iOS simulator on a iPhone 5 running iOS 8.0. This is with AngularJS v1.3.13.

    [Error] Error: Attempted to assign to readonly property.
    ChildScope@http://localhost:9000/bower_components/angular/angular.js:13682:42
    $new@http://localhost:9000/bower_components/angular/angular.js:13690:40
    boundTranscludeFn@http://localhost:9000/bower_components/angular/angular.js:7092:40
    controllersBoundTransclude@http://localhost:9000/bower_components/angular/angular.js:7756:35
    ngRepeatAction@http://localhost:9000/bower_components/angular/angular.js:24502:26
    $watchCollectionAction@http://localhost:9000/bower_components/angular/angular.js:14110:21
    $digest@http://localhost:9000/bower_components/angular/angular.js:14243:31
    $apply@http://localhost:9000/bower_components/angular/angular.js:14506:31
    done@http://localhost:9000/bower_components/angular/angular.js:9659:53
    completeRequest@http://localhost:9000/bower_components/angular/angular.js:9849:15
    requestLoaded@http://localhost:9000/bower_components/angular/angular.js:9790:24
      (anonymous function) (angular.js, line 11607)
      (anonymous function) (angular.js, line 8557)
      $digest (angular.js, line 14261)
      $apply (angular.js, line 14506)
      done (angular.js, line 9659)
      completeRequest (angular.js, line 9849)
      requestLoaded (angular.js, line 9790)

**Comment by caitp** [comment]
@pascalc great, now provide a reproduction =)

**Comment by chriswu14** [comment]
Using AngularJS v1.2.22
    Error: Attempted to assign to readonly property.
    $$childScopeClass@http://XXX/bower_components/angular/angular.js:12037:23
    $new@http://12wbt-frontend.dev/bower_components/angular/angular.js:12045:45
    compositeLinkFn@http://XXX/bower_components/angular/angular.js:6067:38
    compositeLinkFn@http://XXX/bower_components/angular/angular.js:6089:24
    compositeLinkFn@http://XXX/bower_components/angular/angular.js:6089:24
    compositeLinkFn@http://XXX/bower_components/angular/angular.js:6089:24
    nodeLinkFn@http://XXX/bower_components/angular/angular.js:6686:35
    compositeLinkFn@http://XXX/bower_components/angular/angular.js:6086:23
    publicLinkFn@http://XXX/bower_components/angular/angular.js:5982:45
    http://XXX/assets/scripts/app.js:51874:49

and the code in app.js is '$compile(element.contents())($scope);' . The error is reproduced on Ipad safari with IOS Simulator. Commenting out 'user strict' does fix the error as well.

**Comment by caitp** [comment]
@chriswu14 please provide a reproduction in addition, thanks!

**Comment by olivermuc** [comment]
Seeing the issue as well. That said, not sure how successful or even useful the 'reproduction' request is, in light of the fact, that this phenomena seems *very* specific to certain setups. Eg. I only ran into this with **ui-router**, **child ui-states & views** AND **the iphone 5 iOS simulator**. 

Remove any of the variables and it works fine, with the key issue apparently being the iPhone5 sim.

The iPhone 4S and 5S work like a charm and so does the iPad2 / retina sim.

**Comment by m3kka** [comment]
I agree. More than one person tried to reproduce this in a controlled environment unsuccessfully (me included), I guess it won't be fixed anytime soon.

**Comment by leonprou** [comment]
I experience the same problem on iPhone 6 also, developing with Ionic
framework

**Comment by promactkaran** [comment]
I am too having the same issue iphone 6 simulator running iOS 8.2

**Comment by Karl-Gustav** [comment]
I'm having the same problem on my iPad (3rd generation) running iOS 8.1.3 (12B466), with remote logging. And the problem goes away when I remove "use strict" from the angular.js source code. Have anybody notified Apple about this?

**Comment by bennett000** [comment]
Without a reproduction it's unlikely this issue can be taken seriously by anyone, and those (relatively) few of us affected by the issue have mostly found alternate solutions.  Every so often I set aside some time to attempt a reproduction, and I get nowhere.

I've experienced the problem on all versions of iOS >= 8.0.0, but only on a few models.  I've tried several Angular versions in the 1.2.x, and 1.3.x lines, up to 1.3.14 - I don't encounter the issue in any of my unit tests, or e2e component tests - only in one particular application, typically on route change, and only after the app has been running for a while.

jhunken 's solution has worked very well for me.  Using it _unminified_ I don't have issues at all.  However minifications with Uglify, and Closure Compiler break the fix.  I have not had time to try excluding the decorator from the minification process, and catting it on yet...

**Comment by caitp** [comment]
just fwiw, it's okay if the reproduction is not "minimal", though simpler is better --- you can link to an affected app (ideally with minimal backend requirements, so it's easier for us to build a repro with)

**Comment by iainbeeston** [comment]
@caitp I'm seeing this consistently [here](https://brojure.com/brojure/seawolf/a/running-upper-deck). If you click the "cross" in the upper-right of the screen, I get the error described (and the main part of the screen goes white - it should navigate to another page). Note that it doesn't happen in the iOS simulator, but using iOS 8.2 on an iPhone 5 I can reproduce every time.

**Comment by caitp** [comment]
@iainbeeston I can't repro that on my iphone 6 or ipad air :( I'll see what I can do though

**Comment by iainbeeston** [comment]
I've cleared the cache and it still happens on my iPhone 5.

I've also just tested on an iPhone 6 (also iOS 8.2) and as you said, it works correctly there. So it seems to be specific to iPhone 5x (not sure about iPhone 4x).

**Comment by darrenmerrett** [comment]
I collect Angular JS errors from my website and I see this error on random iPads and iPhones based on user agent.

I have tried to reproduce this on an iPhone 5, iPad mini2 and iPad air and I cannot replicate it.

**Comment by bennett000** [comment]
@iainbeeston I tried that on my iPad but I wasn't 100% sure what you meant by cross - can I assume the layout is different on the pad? 

Anyway clicking on a few things, including the toggling menu, and an 'x' didn't throw any errors on my pad, IIRC my mini's model is MD528C/A

**Comment by iainbeeston** [comment]
It's possible it only affects iphone 5 and before.

**Comment by caitp** [comment]
I've asked a family member with an iphone 5 (no additional details on the model #, but hey) on iOS 8.2 to try it, and they sent back screenshots that appear to be working just fine. This seems like a bit of a heisenbug, maybe it's worth reporting it as a bug on Safari itself

**Comment by iainbeeston** [comment]
Hey that's interesting. Works for me too now (we've updated the code since
then, so perhaps an unrelated change fixed it?)

Does anyone else have a site that can reliably reproduce the issue?

**Comment by khashayar** [comment]
I have a reproducible scenario on the website I'm working on, but I can only share the login details privately. Please give me your email so I can send you the details.

**Comment by iainbeeston** [comment]
@caitp I'm still getting the problem actually. If you go to [the same page as before](https://brojure.com/brojure/seawolf/photos), then click the arrow at the top-left arrow (which loads the previous page), then the arrow at the top-right (which goes back to the original page) I get the error fairly reliably.

![photo 02-04-2015 3 08 25 pm](https://cloud.githubusercontent.com/assets/111963/6965570/84b9498c-d94a-11e4-965a-8f2e5fc6d90f.png)

If it doesn't happen immediately, try clicking back and forth a few times (just click on the arrows, but not the cross icon). You can tell the error has occurred because the image no longer appears.

![photo 02-04-2015 3 13 29 pm](https://cloud.githubusercontent.com/assets/111963/6965621/f8fca654-d94a-11e4-98c7-b833879a406a.png)

**Comment by iainbeeston** [comment]
This is on an [iPhone 5 (A1429)](https://support.apple.com/kb/SP655?locale=en_US), running iOS 8.2 and mobile safari 8.0 (useragent `Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12D508 Safari/600.1.4`)

**Comment by caitp** [comment]
as explained, I'm unable to reproduce this on 3 different iOS devices, or any of the iOS simulator devices.

I think you should file this as a bug on webkit or contact Apple. There is very little anyone can do if we can't see the bug ourselves

**Comment by iainbeeston** [comment]
Sorry, I just figured that a more detailed explanation would help reproduce the bug.

Is it worth closing this issue then?

**Comment by darrenmerrett** [comment]
It seems to me this bug is random on devices and iOS versions. I have several devices and I cannot reproduce the fault, but I know that some people on my website do experience this problem because I see the error messages.

I would suggest that under these circumstances, perhaps it's something about the configuration of the devices, some settings or perhaps some other software that's been installed that causes the conflict.

**Comment by asbjornu** [comment]
This is obviously a problem for a lot of people. The fact that it is not reproducible in a controlled environment does in my opinion not preclude any fixes for the given problem to be implemented. As it seems, there is a [monkey-patch](https://gist.github.com/jhunken/767cdf62ecc9f31c56da) available that fixes the problem. If said monkey-patch does not cause any other verifiable problems (in unit tests etc.) I don't really understand the reasoning behind not applying it and fixing the problem for everyone.

**Comment by caitp** [comment]
> If said monkey-patch does not cause any other verifiable problems (in unit tests etc.) I don't really understand the reasoning behind not applying it and fixing the problem for everyone.

send a PR =)

**Comment by asbjornu** [comment]
@caitp wrote:
> send a PR =)
I'd prefer if @jhunken as the author of said monkey patch could do it.

**Comment by bennett000** [comment]
@asbjornu I've been lucky enough to benefit from jhunken's patch *but* in my use case the monkey patch has not worked after minification.  If my problem is the same problem other people have; accepting this patch won't necessarily fix anything.

If I can get the go ahead to publish our demo, then we should be able to reproduce on other devices.  This might not be for another month though.  There is also the chance that the recent iOS 8.3 release has fixed this problem, but I haven't had a chance to test yet.

**Comment by Totemika** [comment]
Hi, @spencerwi , have you figured out why the "undefined" word appears when applying the patch? Have you found any other workaround? Thanks

**Comment by brauliodiez** [comment]
Update about the "undefined", in our case this was coming from the angular translate (translate directive):

<li><a id="headingHome" (...) translate="HOME"></a></li>

we have moved to controller code and we don't get the "undefined" (just some rough code).

$('#headinghome')[0].innerText= $translate.instant('HOME');

**Comment by brauliodiez** [comment]
@totemika @spencerwi from our side we just get rid of the "undefined" by changing the way we are using translate, e.g.

it was not working using:
```html
<h2 translate="'GETTING_STARTED_HEADLINE"></h2>
```
It works with:
```html
<h2>{{ 'GETTING_STARTED_HEADLINE' | translate }}</h2>
```

**Comment by Totemika** [comment]
@brauliodiez @spencerwi That totally worked :+1:

**Comment by maximumduncan** [comment]
@bennett000 btw - I've just upgraded to iOS 8.3 and I can confirm that I'm still getting this issue.

**Comment by brauliodiez** [comment]
Are you using angular translate?

**Comment by maximumduncan** [comment]
@brauliodiez - I am, yes.

**Comment by bewithjonam** [comment]
@pascalc On removing 'use strict', I also see the browser hanging after 5-10 mins. 
angularjs-1.3.15, ios-8.3, iPad-2, ui-router-0.2.13

Any other workarounds that solved this problem for you? Any help would be much appreciated.

Webkit issue for the same https://bugs.webkit.org/show_bug.cgi?id=49739

**Comment by gitmh** [comment]
I use angular material and getting the same "Error: Attempted to assign to readonly property." almost always on my 8.3 iOS iPad (first retina). Monkey patch does not solve the problem for me. Also deleting "use strict" in angular and angular-material makes everything worse. There are no error messages in safari remote debugger anymore but the page loads only in parts - from ng-repeat half of the items are randomly displayed correct, other half is shown with {{...}} as plain text.

Maybe somebody can use my testcase http://v2.atocc.de/testcase just click on "öffnen" of one of the listed diagrams on an iPad - I always get 4x"Error: Attempted to assign to readonly property." in about 80% of my many reloads / tries - but still sometimes it works. The testcase works fine on desktop browsers. I have another angular-material app without any problem so I can't simplify the testcase.

**Comment by caitp** [comment]
@gitmh thanks for the reproduction, taking a look on my ipad =)

This has been extremely frustrating to track down, I'm sorry for everyone having issues supporting iOS devices due to this problem :(

**Comment by caitp** [comment]
@gitmh I've been trying for a while, but I can't seem to reproduce this on an iPad (MD787C/A) with iOS 8.3 --- Method: Press on "öffnen" buttons on random graphs (whether same button is pressed sequentially or not doesn't seem to have an impact).

If any WebKit/JSC gurus know if there are any flags to force (different types of) optimization of even cold code, that might be helpful in tracking this down.

**Comment by joshuajohnson814** [comment]
@gitmh I'm getting the "Attempted to assign to readonly property" error about every other time I click on offnen with iPad 2 simulator iOS 8.3 (12F69). 

I'm having the same issue on my Angular application :(

**Comment by gitmh** [comment]
For me this looks like a timing issue on the iPad - my example is loading a diagram with multiple small parts with lot's of bindings. A small diagram works better - lets say 30% of the time and a large one only in 5% of my tries. The result for the user is an empty diagram on the right side of the window. The ng-repeat div is empty because it breaks on the error. Seldom also the starting page of the testcase only shows some cards with "öffnen" buttons - so material buttons seems to be a good example for reproduction. I made another copy http://v2.atocc.de/testcase2 where I deleted "use strict" in angluar, angular-route, angular-material and angular-aria. This makes the whole app unuseable but no error messages.

**Comment by gdjennings** [comment]
@gitmh I agree, timing/load seems to be the trigger. In my application, I don't see this error until I stress the CPU. Other conditions mentioned in this thread (iOS version, translate, etc) do not seem to be related as I have seen it on iPad mini, iPad air 2, iPhone 5s, and the simulator on every 8.x version of iOS and also in cordova using UIWebView and WKWebView. I have no log of it happening on any device since removing 'use strict'. It also explains why it is so damn hard to reproduce.

**Comment by gitmh** [comment]
I added another testcase http://v2.atocc.de/testcase5 with the monkey patch from @jhunken it works much better but still sometimes the app breaks without error message. The "Error: Attempted to assign to readonly property." messages disappeared now.

**Comment by oori** [comment]
+1 Bumped into the same as everyone describe.

1. I can confirm the error using @gitmh testcase on my test iPad MD328FD/A with iOS 8.3
2. My app is consistently hitting this issue. but it's in private-beta, shareable only via PM.
3. I use some of the same dependencies others mention:
angular#1.3.15, angular-ui-router#0.2.14, angular-translate#2.4.2 
4. Removing "use strict" makes safari hang as @pascalc and @bewithjonam mention

**Comment by adrienkohlbecker** [comment]
+1 also, I'm getting this on multiple iphones/ipads all running some version of iOS 8
2 hours before the first occurence was reported, I merged this in my bower.json (no issue whatsoever before this merge, and no other merge were done that day)

```
-    "ngCordova": "0.1.14-alpha",
-    "ionic": "1.0.0-rc.3",
-    "mapbox.js": "2.1.6",
+    "ngCordova": "0.1.15-alpha",
+    "ionic": "1.0.0-rc.4",
+    "mapbox.js": "2.1.8",
```

Here are the full dependencies:
```
  "dependencies": {
    "ngCordova": "0.1.15-alpha",
    "ionic": "1.0.0-rc.4",
    "angular-animate": "1.3.15",
    "angular-sanitize": "1.3.15",
    "mapbox.js": "2.1.8",
    "leaflet": "0.7.3",
    "rollbar.js": "1.2.1"
  },
  "resolutions": {
    "angular": "1.3.15",
    "angular-animate": "1.3.15",
    "angular-sanitize": "1.3.15"
  }
```

**Comment by spencerwi** [comment]
In update to our stack, we've been running without `"use strict"` for a while, and have started to hit Safari lockups. In the worst cases, users have to power off and power on the device to get Safari responsive again.

I'm optimistic now that there's a consistent reproduction that a fix might be possible, but in the short term, has anyone found a viable workaround?

**Comment by Cameri** [comment]
+1

**Comment by SonicWizard** [comment]
I see the problem too on non-64 bit devices running iOS 8. iPad 2/3/4, mini, etc. are affected. Devices NOT affected are iPad Air/Air 2, mini 2/3.

It is an Apple bug, but since other frameworks (Ember, Require, etc.) are putting in workarounds, it'd be nice to see Angular implement a workaround as well.

btw, can anyone comment on where to submit these kinds of bugs to Apple?

**Comment by SonicWizard** [comment]
Also, I'll add that the removal of 'use strict' only goes so far. Like @spencerwi stated, we see Safari lock up completely and user has to force quit.

**Comment by SonicWizard** [comment]
@HAKASHUN and @m3kka 's solution of changing from dot to array notation worked. Why 32-bit Safari is choking on dot notation on those lines, I don't know. That's a mystery Apple probably needs to solve.

**Comment by SonicWizard** [comment]
Can the array notation be put into Angular proper to resolve this problem for everyone?

**Comment by aaronharnly** [comment]
We (Amplify Education, http://amplify.com) are also encountering this issue. We tried removing 'use strict', various changes in our stack. Currently considering using this patch to work around it.

**Comment by zourtney** [comment]
By the way, I've had success making @jhunken's `decorator` patch minification-safe by declaring the offending property names as string `var`s, then referencing them with bracket notation.

    var $id = '$id',
        $$childScopeClass = '$$childScopeClass',
        ...

And then:

    ...
    this[$id] = nextUid();
    this[$$childScopeClass] = null;
    ...

Full forked [gist available here](https://gist.github.com/zourtney/dcd8914d68b31a4f1f39) (modeled off Angular 1.2.23)

**Comment by harrastia** [comment]
@zourtney thanks for publishing that, I added the decorator to our app which was producing quite many tracebacks related to this from iOS devices. It has been a life saver as our app was totally unusable, especially on iPads. We are currently using Angular 1.3.8.

Seems there are more places where there is this misbehaviour for iOS devices, we got new tracebacks in Sentry now pointing to this line:

https://github.com/angular/angular.js/blob/bab474aa8b146f6732857c3af1a8b3b010fda8b0/src/ng/compile.js#L1096

```
this.$attr[key] = attrName = snake_case(key, '-');
```

most likely this will work if written in same format as in your patch

```
var $attr = '$attr';
this[$attr][key] = attrName = snake_case(key, '-');
```
This is aligned with the comments in this thread around Dec 9 2014 https://github.com/angular/angular.js/issues/9128#issuecomment-66314752

Unfortunately I don't have time in the coming days to try setting up a reproducible environment for this.
Some more info to create a reproducible environment: this is triggered in a "non interactive" call as the starting point of the traceback is here https://github.com/angular/angular.js/blob/v1.3.8/src/ng/browser.js#L49

**Comment by harrastia** [comment]
I forked and adapted the gist decorator by @zourtney to work in Angular 1.3.8 here: https://gist.github.com/harrastia/4a963e62605f73bbfff2

**Comment by Crash--** [comment]
To follow the webkit issue : https://bugs.webkit.org/show_bug.cgi?id=138038. Seems that a patch is on the way.

**Comment by sevinlim** [comment]
Not sure if it helps but for me I came into the problem when using a tab view in another ion-view which can be taken over by other views. Navigating back to the tab states will give me the error you guys have been having. My fix was to use 2 $state.go. First one navigating to the 'tab' state and wrapping the other in a $timeout function with the tab state that i wanted to go to.

**Comment by erwingaitano** [comment]
+1
Removing 'strict mode' from angular.js just kind of alleviate the problem but it still happens randomly.

bower.json:

    "dependencies": {
        "angular": "1.3.15",
        "angular-route": "~1.3.5",
     }

**Comment by rebelliard** [comment]
I just encountered this problem on my production site, and I noticed that the bug came from my Grunt's **HTML minification** task (really!), and specifically, from having the [**removeOptionalTags**](http://perfectionkills.com/experimenting-with-html-minifier/#remove_optional_tags) option enabled.

The second I disabled `removeOptionalTags`, everything started working again.

**Comment by cfey** [comment]
removing 'use strict' worked for me too, using iphone 6 emulator
worked on android perfectly, even without removing use strict!

for everyone who uses ionic it is in the line 8705
haven't experienced any problems so far

**Comment by JobaDiniz** [comment]
I ran into this today as well for my website.. I'm using Angular 1.2.28.

**Comment by hcentelles** [comment]
:+1: on ios 8 emulator, ionic 1.0, angular 1.3.16

**Comment by nuvorm** [comment]
I am only experience this bug when having form elements wrapped in a directive template.
With or without a transclude.

currently i have a contact form directive. template html is:
```javascript
<div>
<container h-max="1">
  <column align="left">
    <ng-transclude></ng-transclude>
  </column>
</container>
</div>
```

and in the index file i use it:

```javascript
<contact-form css-prefix="contact-form">
   
     <ng-input form-id="voornaam" class="form-control"></ng-input>
     <ng-input form-id="achternaam" class="form-control"></ng-input>
     <ng-input form-id="bedrijf" class="form-control"></ng-input>
     <ng-input form-id="telefoon" class="form-control"></ng-input>
     <ng-input form-id="email" class="form-control"></ng-input>
     <ng-text-area form-id="vraag" class="form-control"></ng-text-area>
     <button ng-click="sendContactForm()" class="btn btn-primary contact-form-button"><lang parent-id="contact_form" id="button"></lang></button>
    </contact-form>
```

There is much more code in my project but i was wondering if it could be due to form elements causing this bug.

**Comment by dhyegofernando** [comment]
When will it be applied in the master branch? Any idea?

**Comment by oori** [comment]
@dhyegofernando you should ask apple...   the bug is fixed,  but isn't deployed. see here:   https://bugs.webkit.org/show_bug.cgi?id=138038#c19

**Comment by johgusta** [comment]
We've been using @jhunken's decorator but ran into issues with  scope id's when upgrading to Angular 1.4. We managed to work around it by wrapping the $new function in a try catch block, inspired by this thread: 
http://stackoverflow.com/questions/13147026/disabling-jit-in-safari-6-to-workaround-severe-javascript-jit-bugs

This updated monkey patch decorator works for us:
https://gist.github.com/johgusta/46fb1ecd459dbc3bfdfc

**Comment by jacksonrayhamilton** [comment]
My company has released a utility which rewrites our code with all dot property assignments changed to bracket notation assignments.

That is, it rewrites the following code

```js
object.$$a = 5;
```

to this

```js
var __webkitAssign__$$a = '$$a';
object[__webkitAssign__$$a] = 5;
```

as a workaround to prevent a `TypeError`. [The utility is available on npm.](https://www.npmjs.com/package/webkit-assign) It has alleviated the "readonly" issue for us on iOS.

For a while, my company simply removed `'use strict';`, but recently we encountered a strange bug where repeatedly opening dialogs from angular's material design library resulted in the page freezing. An error was thrown: `Error: null is not an object (evaluating 'parent.$$childTail.$$nextSibling = child')`, and I suppose that somehow resulted in `$digest` looping infinitely (thus the freeze).

We tried adding `'use strict';` back, but then we started seeing the "readonly" error at exactly the same time as the aforementioned error. This leads me to believe the `null` reference is related to this JIT bug, and that the bug is not (or hopefully, *was* not) limited to just strict mode.

The utility isn't perfect, but it at least seems to fix the issue on the `angular.js` distributed via bower.

**Comment by LoganLehman** [comment]
I have verified that @jacksonrayhamilton 's solution works wonderfully on v1.2.26

**Comment by EikeLueders** [comment]
For me @jacksonrayhamilton 's solution also works flawlessly on v1.4.1

**Comment by Nivani** [comment]
I mentioned before that patching AngularJS worked for me, in a new project it was not enough. I used the utility that @jacksonrayhamilton posted and now it works, thanks! (v1.3.16)

**Comment by lgalfaso** [comment]
This was finally fixed in webkit https://bugs.webkit.org/show_bug.cgi?id=138038#c21 but it will take some time for the updates to be on everybody's phone

**Comment by iliyanwhich** [comment]
I am using this patch and it created another issue on angularjs 1.4.3 (previously I used 1.2.28):
If you have directive with isolated scope that contains ngRepeat (or another ng directive that creates new scope) it won't evaluate any expressions inside the ngRepeat HTML. It will just print "text with {{ variableName }}" instead of the value.
Here is the issue:
http://stackoverflow.com/questions/31919921/angularjs-1-4-3-ngrepeat-prints-variablename-instead-of-value-when-used-in

**Comment by onesoftdk** [comment]
We are using:

```
{
    ...
    "angular": "1.3.1",
    "angular-translate": "2.4.2",
    "ui-route": "0.2.15",
    ...
}
```

And as @brauliodiez  described we just got rid of the "undefined" by changing the way we are using translate,

from not working using:
```
<h1 translate>resetPassword.ChangePassword</h1>
```
to working with:
```
<h1>{{::'resetPassword.ChangePassword' | translate}}</h1>
```
We haven't used any monkey patch or hasn't removed `'use strict';`

**Comment by cgewecke** [comment]
+1 on efficacy of @jacksonrayhamilton solution using Angular 1.3.16. Can also confirm that this problem basically breaks apps that use Angular Material.

**Comment by albertogasparin** [comment]
@johgusta [solution](https://gist.github.com/johgusta/46fb1ecd459dbc3bfdfc) worked really well on our website, plus it looks much safer as it is not rewriting Angular internals.

**Comment by zixia** [comment]
I ran into this issue yesterday, got "readonly property" error. but it's very strange that this bug is not always appear. I use the same simulater & sourcecode in same mac air notebook, sometimes got error, sometimes no error at all.

**Comment by wesjones** [comment]
Currently getting this error on iOS 8.4.1, angular 1.4.3. Only getting this on iPad mini.

**Comment by alscheuring** [comment]
@jacksonrayhamilton I went to try the utility but I'm a little clear on "create a transformed angular.webkitassign.js in the same directory". I'm not sure what I do there.

**Comment by jacksonrayhamilton** [comment]
@alscheuring, in the future please file issues for webkit-assign on our [Issues page](https://github.com/StoryCloud/webkit-assign/issues).

webkit-assign provides a command-line interface. Open a terminal and do `npm install -g webkit-assign` to install the program. `cd` to the directory of a file you'd like to transform. Then type `webkit-assign FILENAME.js` (where "FILENAME.js" should be replaced with an actual filename), and the program will produce a fixed copy of the file in the same directory.

**Comment by ronaldcurtis** [comment]
+1 to @jacksonrayhamilton's solution. I have used @jhunken's patch but it seemed to only reduce the frequency in which the error occurs (still seems to happen randomly).

**Comment by Cinamonas** [comment]
@jacksonrayhamilton's solution is great. We're transforming angular.js with a custom webpack loader:

```js
var webkitAssign = require('webkit-assign');

module.exports = function (source) {
  this.cacheable();
  return webkitAssign(source);
};
```

**Comment by wombleton** [comment]
I published @Cinamonas' loader: https://www.npmjs.com/package/webkit-assign-loader

**Comment by skusunam** [comment]
@johgusta @albertogasparin Which version of Angular are you guys using to claim that this solution (https://gist.github.com/johgusta/46fb1ecd459dbc3bfdfc) is working? This fix doesn't work for us with Angular-1.4.6.

**Comment by brunofranco** [comment]
@johgusta's solution solved my problem in angular 1.4.2

**Comment by illumimari** [comment]
@johgusta's solution appears to have solved our issues in angular 1.4.5

**Comment by shershen08** [comment]
+1, tried with 1.4.0, 1.4.6 same story. How I can intelligently patch Angular if I don't use webpack?

**Comment by lgalfaso** [comment]
@shershen08 https://www.npmjs.com/package/webkit-assign

**Comment by code-tree** [comment]
Note, while reports are that it only affects IOS8, I now have a case of it occurring on IOS9.2 (despite using webkit-assign fix). Any ideas?

> TypeError: Attempted to assign to readonly property.
> Mozilla/5.0 (iPhone; CPU iPhone OS 9_2 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13C75 Safari/601.1

Will try removing 'use strict's once I have a device I can regularly test on.

**Comment by jacksonrayhamilton** [comment]
@code-tree If you've run all JavaScript code on your site through webkit-assign and can still consistently reproduce that error, you should [file a bug](https://github.com/StoryCloud/webkit-assign/issues/new) on our repository.

**Comment by astopo** [comment]
+1, angular 1.3.6 & 1.4.8, ios 8.2

**Comment by code-tree** [comment]
@jacksonrayhamilton Retrying on a different friend's device running iOS 9.2 worked fine, so I can't reproduce it, but it did still occur on the device mentioned earlier (which I no longer have access to). So I guess I'll wait and see if anyone else reports anything.

**Comment by JeroenNelen** [comment]
Hello everyone,

I have also, unfortunately, run into this issue. 
I'm still a bit dazzled because of the behaviour:

- I do not get it on every page in my application
- I do not get it all the time on that specific page, only about 50% of the time

Does anyone noticed that it only happens with specific code or with loading issues?

I will look into the solution @jacksonrayhamilton provided, allthough I'm not sure if this will be usable for our production build.

Kind regards,

**Comment by code-tree** [comment]
Yes, it first occurred for me **after** testing on a device for 30 minutes without any issues. All of a sudden it kept occurring on the index page, without anything being different to the previous 30 minutes...

**Comment by JeroenNelen** [comment]
Hello everyone,
I have noticed that it for some reason in my case seems to be related to the usage of ng-if's. 
I use ng-if's to hide my pages until I have all my back-end calls done in order to make sure that all the data used by the directives is available before loading them. 
Removing these ng-if's, it seems that I no longer get the readOnly errors (tested about 20 times). 
Did anyone else noticed this?

Are there any negative side-effects for using @jacksonrayhamilton his solution? I am currently using angular 1.4.7.

**Comment by jacksonrayhamilton** [comment]
@JeroenNelen While it's possible `ngIf` could be a culprit, I think it's more likely that the `ngIf` prevented or triggered some other code that was a culprit. As for the efficacy of the solution, we use an AST to transform and fix the code, so it should be pretty precise. We haven't had any problems using it in our production build and we have no open issues currently.

**Comment by JeroenNelen** [comment]
@jacksonrayhamilton Thanks for your reply. I will test to see if the issue still occurs after these changes, if it indeed occurs, I think the only solution will be to use your package (nice job on it by the way) :). 
The weird thing is that I cannot get the error to occur on any other page, only on these specific pages, therefor I was meddling with the code on those pages to see if I could prevent the error locally (rather than adjusting anguluar.js).

**Comment by quochuy** [comment]
@jacksonrayhamilton Hi, is there a way to use your tool with Grunt?

**Comment by jacksonrayhamilton** [comment]
```js
var webkitAssign = require('webkit-assign');
module.exports = function(grunt) {
  grunt.registerTask('webkitAssign', function() {
    var file = 'build/scripts.js';
    var code = grunt.file.read(file);
    var transformed = webkitAssign(code);
    grunt.file.write(file, transformed);
  });
};
```

Or for better integration, maybe you'd like to submit a plugin to our repository? :-)

**Comment by spenoir** [comment]
We found a bug with angular-translate and ng-show. removing ng-show in favour of ng-if fixes the issue

**Comment by lesleychard** [comment]
@jacksonrayhamilton Thank you! Utility worked flawlessly. Problem is no longer persisting in iOS 8.3 / iPhone 5, AngularJS 1.4.9. Hours of headache later, you have made my day.

**Comment by jakub-g** [comment]
@spenoir I think you're talking about a different bug regarding `ng-show` / `ng-hide`:
https://github.com/angular/angular.js/issues/13380#issuecomment-181309484

**Comment by esetnik** [comment]
I just encountered the same issue. Will try the webkit-assign fix and report back.

**Comment by gkalpak** [comment]
Since this is a bug in WebKit, which [has been fixed][1], there is a [work-around][2] and we don't have a clear reproduction to investigate further, this is not actionable on our part.
Closing...

[1]: https://github.com/angular/angular.js/issues/9128#issuecomment-125735846
[2]: https://www.npmjs.com/package/webkit-assign

**Comment by celle886** [comment]
@code-tree i ran into the same issue with the same description (iOS 9, appearing after long usage time). How do you fix your app?

**Comment by code-tree** [comment]
@celle886 No idea, sorry :)

It seems like webkit-assign helps, but I couldn't reliably reproduce the problem.

**Comment by mpiasta-ca** [comment]
Just came across this issue in our web app with Angular 1.4.8 when using iPod/iPhone 8.4.1 on Safari 8.0. Thanks for whoever was first to suggest webkit-assign. We used webkit-assign/gulp and that solved a problem that has been ghosting/plaguing us for 6+ months.

**Comment by TiagoSilvaPereira** [comment]
I have this issue using Angular 2 (with Ionic 2) in the iPhone 6s / iOS 9.2 simulator. The message is "Exception: TypeError: Attempted to assign to readonly property" I've tried the @jacksonrayhamilton webkit-assign solution, but this does not seem to work with Angular 2 (I've tried in the final app.bundle.js file).

Someone have an ideia how to fix this problem?

Thanks!

**Comment by Narretz** [comment]
@TiagoSilvaPereira Please open an issue about this on the https://github.com/angular/angular repo. This repo here is only for the 1.x branch

**Comment by TiagoSilvaPereira** [comment]
Thanks @Narretsz, i have solved this problem removing all "use strinct" from my code.
